Add Redis task handler

michalc opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

This stores log lines in Redis up to a configured maximum log lines, always keeping the most recent, up to a configured TTL.

This deviates from other existing task handlers in that it accepts a connection ID. This allows it to be used in addition to other handlers, and so allows a graceful/reversible transition from one logging system to another.

This is particularly useful in situations that use Redis as a message broker, where additional infrastructure isn't desired.

closes: https://github.com/apache/airflow/issues/31834

michalc committed:
Add Redis task handler

This stores log lines in Redis up to a configured maximum log lines, always
keeping the most recent, up to a configured TTL.

This deviates from other existing task handlers in that it accepts a connection
ID. This allows it to be used in addition to other handlers, and so allows a
graceful/reversible transition from one logging system to another.

This is particularly useful in situations that use Redis as a message broker,
where additional infrastructure isn't desired.

closes: https://github.com/apache/airflow/issues/31834

utkarsharma2 commented on airflow/providers/redis/log/redis_task_handler.py:
```diff
@@ -0,0 +1,105 @@
+        If omitted, this is the equivalent of 28 days.
+    :param conn_id:
+        Airflow connection ID for the Redis hook to use
+        If omitted or None, conf.get("logging", "REMOTE_LOG_CONN_ID") is used.
```
`conf.get("logging", "REMOTE_LOG_CONN_ID")` is this supposed to be here? 
If you intend to specify that airflow's logging config will be used, would be better to mention that explicitly. This probably won't make sense to users, unless they are aware of internals.

michalc replied:
Have changed it to `logging.remote_log_conn_id`, which is how the documentation refers to the setting at https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#logging

utkarsharma2 commented on airflow/providers/redis/log/redis_task_handler.py:
```diff
@@ -0,0 +1,105 @@
+    ):
+        log_str = b"\n".join(self.conn.lrange(self._render_filename(ti, try_number), start=0, end=-1)).decode(
+            "utf-8"
+        )
```
```suggestion
        log_str = b"\n".join(self.conn.lrange(self._render_filename(ti, try_number), start=0, end=-1)).decode()
```
default encoding format is `utf-8` ref - https://docs.python.org/3/library/stdtypes.html#bytes.decode

michalc replied:
Now removed

utkarsharma2 commented on airflow/utils/log/file_task_handler.py:
```diff
@@ -147,7 +147,7 @@ class FileTaskHandler(logging.Handler):
     def __init__(self, base_log_folder: str, filename_template: str | None = None):
         super().__init__()
-        self.handler: logging.FileHandler | None = None
+        self.handler: logging.Handler | None = None
```
This change look out of place?

michalc replied:
This is true - but it allows the lower level handler, `_RedisHandler` to not inherit from `logging.FileHandler`, which I think is appropriate since there is no file involved(?)

It looks like the codebase already doesn't enforce that the lower level handler extends from `logging.FileHandler` - https://github.com/apache/airflow/blob/2940b9fa55a6a72c60c2162e541631addec3d6b8/airflow/providers/amazon/aws/log/cloudwatch_task_handler.py#L67 which is defined in https://github.com/kislyuk/watchtower/blob/698ff2241befb6586269b6bbce0e305032c2d45c/watchtower/__init__.py#L139

This touches on the bits that I'm unsure about - it looks like `FileTaskHandler` has a lot of file-related things, but also it looks like it's expected to extend from it when not logging to a file. Not quite sure how to tackle that

utkarsharma2 replied:
Should have been more specific initially. :)

What I meant was that maybe this change is not required by this PR? since you have anyway overridden the         `self.handler` in class RedisTaskHandler. Also, `FileTaskHandler` is used in many places, and changing it here might affect those implementations. 

About inheriting from `FileTaskHandler` I'm not sure either, maybe @bolkedebruin might have more context.

michalc replied:
> What I meant was that maybe this change is not required by this PR? since you have anyway overridden the self.handler in class RedisTaskHandler

Ah - so if I change it back to how it was before this PR, I get a type error:
```
airflow/providers/redis/log/redis_task_handler.py:61: error: Incompatible types
in assignment (expression has type "Optional[_RedisHandler]", base class
"FileTaskHandler" defined the type as "Optional[FileHandler]")  [assignment]
```
So something else would have to change too. The easiest way that I've come up with is to de-facto disable type checking on `self.handler` by removing type annotations in a few places, to make it more like the Cloudwatch handler, but that seems a bit... cheat-y.

`_RedisHandler` could really inherit from `logging.FileHandler` to satisfy type checking, but its constructor seems to require a path to a file, which seems odd in this case.

uranusjr replied:
This change is OK to me since the base class only really needs to inner handler to be a `logging.Handler`. It could be moved to a separate PR but since this is where the change becomes necessary (instead of just a generally good thing to do) it’s fine to be included here.

eladkal replied:
If this specific change is not tightly coupled to this PR I would suggest to do it in a separated PR.
this is because utils/log goes to core release cycle while the other files goes to provider release cycle.
@michalc Is it possible to separate?

eladkal replied:
@michalc can you check my above comment?

michalc replied:
@eladkal Sounds reasonable to separate to another PR. Will do that now

michalc replied:
@eladkal Done in https://github.com/apache/airflow/pull/32773

hussein-awala reviewed (approved):
Looks good! Waiting the CI

potiuk reviewed (approved):
Nice one!

michalc committed:
Allow FileTaskHandler to delegate to instances of logging.Handler

FileTaskHandler is the base class for logging handlers, including those that
don't log to files via delegating to logging.FileHandler, e.g. in the
CloudwatchTaskHandler at
https://github.com/apache/airflow/blob/2940b9fa55a6a72c60c2162e541631addec3d6b8/airflow/providers/amazon/aws/log/cloudwatch_task_handler.py#L67

It is suspected that type checking is not enabled in this part of the
CloudwatchTaskHandler otherwise it would have already been failing.

This change adjusts the base class so if type checking is enabled in the task
handler, if it delegates to a logging.Handler that is not a logging.FileHandler
as the CloudWatchHandler, then the type checking should pass.

This was originally part of https://github.com/apache/airflow/pull/31855 and
split out.

related: https://github.com/apache/airflow/issues/31834
